experiments/interpretation/dutchf3_patch/local/train.py

```diff
@@ -203,40 +206,43 @@ def run(*options, cfg=None, debug=False):
 
     # Logging:
     trainer.add_event_handler(
-        Events.ITERATION_COMPLETED, logging_handlers.log_training_output(log_interval=config.TRAIN.BATCH_SIZE_PER_GPU),
+        Events.ITERATION_COMPLETED, logging_handlers.log_training_output(log_interval=config.PRINT_FREQ),
     )
     trainer.add_event_handler(Events.EPOCH_COMPLETED, logging_handlers.log_lr(optimizer))
 
-    fname = f"metrics_{config_file_name}_{config.TRAIN.MODEL_DIR}.json" if debug else None
-    evaluator.add_event_handler(
-        Events.EPOCH_COMPLETED,
-        logging_handlers.log_metrics(
-            "Validation results",
-            metrics_dict={
-                "nll": "Avg loss :",
-                "pixacc": "Pixelwise Accuracy :",
-                "mca": "Avg Class Accuracy :",
-                "mIoU": "Avg Class IoU :",
-            },
-            fname=fname,
-        ),
-    )
-
     # Tensorboard and Logging:
     trainer.add_event_handler(Events.ITERATION_COMPLETED, tensorboard_handlers.log_training_output(summary_writer))
     trainer.add_event_handler(Events.ITERATION_COMPLETED, tensorboard_handlers.log_validation_output(summary_writer))
 
+    # add specific logger which also triggers printed metrics on test set
     @trainer.on(Events.EPOCH_COMPLETED)
     def log_training_results(engine):
         evaluator.run(train_loader)
         log_results(engine, evaluator, summary_writer, n_classes, stage="Training")
+        logging_handlers.log_metrics(engine, evaluator, stage="Training")
 
+    # add specific logger which also triggers printed metrics on validation set
     @trainer.on(Events.EPOCH_COMPLETED)
     def log_validation_results(engine):
         evaluator.run(val_loader)
         log_results(engine, evaluator, summary_writer, n_classes, stage="Validation")
+        logging_handlers.log_metrics(engine, evaluator, stage="Validation")
 
-    # Checkpointing:
+    # dump validation set metrics at the very end for debugging purposes
+    if debug:
+        fname = f"metrics_test_{config_file_name}_{config.TRAIN.MODEL_DIR}.json"
+
+        @trainer.on(Events.COMPLETED)
+        def log_final_validation_results(engine):
+            evaluator.run(val_loader)
+            metrics = evaluator.state.metrics
+            out_dict = {x: metrics[x] for x in ["nll", "pixacc", "mca", "mIoU"]}
+            with open(fname, "w") as fid:
+                json.dump(out_dict, fid)
+            log_msg = " ".join(f"{k}: {out_dict[k]}" for k in out_dict.keys())
+            logging.info(log_msg)
```

Comment:
the final validation results will still be logged right? so maybe we re-purpose this to just dump the final results to the disk.

also, is there a way we can avoid evaluating on the validation set twice for the last epoch? Ignite again is complicating this, but it would be nice if we can do it.

Reply:
sure. I actually thought this was cleaner.... but just tied it back up to log_validation_results method